projects/com.oracle.truffle.llvm.nodes/src/com/oracle/truffle/llvm/nodes/intrinsics/c/LLVMTruffleOnlyIntrinsics.java

```diff
@@ -151,70 +152,85 @@ public int executeIntrinsic(LLVMAddress str1, LLVMAddress str2) {
         @Child private ToLLVMNode toLLVM1 = new ToLLVMNode();
         @Child private ToLLVMNode toLLVM2 = new ToLLVMNode();
 
-        @Specialization
-        public int executeIntrinsic(VirtualFrame frame, TruffleObject str1, TruffleObject str2) {
+        @Specialization(limit = "20", guards = {"getSize1(frame, str1) == size1", "getSize2(frame, str2) == size2"})
+        @ExplodeLoop
+        public int executeIntrinsic(VirtualFrame frame, TruffleObject str1, TruffleObject str2, @Cached("getSize1(frame, str1)") long size1, @Cached("getSize2(frame, str2)") long size2) {
             try {
-                return execute(frame, str1, str2);
+                int i;
+                for (i = 0; i < size1; i++) {
+                    Object s1 = ForeignAccess.sendRead(readStr1, frame, str1, i);
+                    char c1 = toLLVM1.convert(frame, s1, char.class);
+                    if (i >= size2) {
+                        return c1;
+                    }
+                    Object s2 = ForeignAccess.sendRead(readStr2, frame, str2, i);
+                    char c2 = toLLVM2.convert(frame, s2, char.class);
+                    if (c1 != c2) {
+                        return c1 - c2;
+                    }
+                }
+                if (i < size2) {
+                    Object s2 = ForeignAccess.sendRead(readStr2, frame, str2, i);
+                    char c2 = toLLVM2.convert(frame, s2, char.class);
+                    return -c2;
+                } else {
+                    return 0;
+                }
             } catch (Exception e) {
                 throw new AssertionError(e);
```

Comment:
I would rather add a CompilerDirectives.transferToInterpreter().
No need to profile the branch; let's cut off this branch from compilation from the start.
This must not happen anyway.